presto-spi/src/main/java/com/facebook/presto/spi/statistics/Estimate.java

```diff
@@ -25,27 +26,38 @@
     //      Skipping for now as there hard to compute it properly and so far we do not have
     //      usecase for that.
 
-    private static final Estimate UNKNOWN = new Estimate(Double.NaN);
+    private static final Estimate UNKNOWN = new Estimate(NaN);
     private static final Estimate ZERO = new Estimate(0);
 
     private final double value;
 
-    public static Estimate unknownValue()
+    public static Estimate unknown()
     {
         return UNKNOWN;
     }
 
-    public static Estimate zeroValue()
+    public static Estimate zero()
     {
         return ZERO;
     }
 
-    public Estimate(double value)
+    public static Estimate of(double value)
+    {
+        if (isNaN(value)) {
+            throw new IllegalArgumentException("value is NaN");
+        }
+        if (isInfinite(value)) {
+            throw new IllegalArgumentException("value is infinite");
+        }
+        return new Estimate(value);
+    }
+
+    private Estimate(double value)
     {
         this.value = value;
     }
 
-    public boolean isValueUnknown()
+    public boolean isUnknown()
```

Comment:
This should be now equivalent to `estimate == UNKNOWN`, right?

Reply:
nan estimate can be a result of nan propagation in calculations in double. I am not convinced disallowing nan is a safer approach  -- code might start to fail in rare cases like when eg certain stats are missing but not the others --  but i don't oppose this change, we can deal with it.

Reply:
Currently `Estimate` class is used only in the SPI. It is not used when estimating statistics in different stats calculators. The point of this change - is to disallow implicit NaN's returned by connector. If the  statistics are unknown - it must be explicitly defined in the code. If there is division by zero that is not supposed to be there - it must be fixed, and not ignored.

Reply:
> Currently Estimate class is used only in the SPI. It is not used when estimating statistics in different stats calculators

i understand. On the other hand Hive connector needs to do some math to produce those stats.

> If there is division by zero that is not supposed to be there - 

This likely will produce inf rather than nan.

> it must be fixed, and not ignored.

as i said, i wouldn't introduce this change, but i am far from vetoing it

Reply:
> This likely will produce inf rather than nan.

http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/Double.java#l69

> i understand. On the other hand Hive connector needs to do some math to produce those stats.

It is really hard to see what math can produce `NaN`, that implicitly means `unknown` in terms of `Estimate` that wouldn't be a programmer's error.